### PR TITLE
improve speed of SparseAutoencoder and make it more flexible

### PR DESCRIPTION
**Pull request by stereomatchingkiss**
This commit intent to complete two things

1 : improve the speed of SparseAutoencoder
I cache the computation result in the data member, so the algorithm do not need to compute it two times.Since the member function is const, I declared the data member as mutable;Do anyone think that make the member function as non-const would be better?

2 : make the SparseAutoencoder more versatile
I add two template parameters as following

template<typename HiddenLayer, typename OutputLayer>
class SparseAutoencoderFunction;

This allow users use different layer from ann two compute the features

### CODE REVIEW

**Comment by rcurtin** [comment]
So you mentioned in IRC that this implementation is faster than the implementation in `src/mlpack/methods/sparse_autoencoder`; can I get a little more information on how you calculated that?  I don't doubt your results or anything, I just want to run some tests of my own and see if I can replicate the speedup on other interesting architectures and systems. :)

File: src/mlpack/tests/sparse_autoencoder_test_2.cpp [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+using namespace arma;
+
+using FSigmoidLayer = ann::SigmoidLayer<ann::LazyLogisticFunction>;
+
+//sparse autoencoder function
+using SAEF = ann::SparseAutoencoderFunction<FSigmoidLayer, FSigmoidLayer>;
+//sparse autoencoder function greedy
+using SAEFG = ann::SparseAutoencoderFunction<FSigmoidLayer, FSigmoidLayer, std::true_type>;
+
+BOOST_AUTO_TEST_SUITE(SparseAutoencoderTest2);
```
Comment (rcurtin) [inline src/mlpack/tests/sparse_autoencoder_test_2.cpp]: Just a note -- when this is merged, we can remove the old `sparse_autoencoder_test.cpp` and revert this to `SparseAutoencoderTest` instead of `SparseAutoencoderTest2`.

File: src/mlpack/methods/ann/sparse_autoencoder_function.hpp [hunk 2]
```diff
@@ -0,0 +1,407 @@
+#ifndef __MLPACK_METHODS_ANN_SPARSE_AUTOENCODER_FUNCTION_HPP
+#define __MLPACK_METHODS_ANN_SPARSE_AUTOENCODER_FUNCTION_HPP
```
Comment (rcurtin) [inline src/mlpack/methods/ann/sparse_autoencoder_function.hpp]: This file looks like it's heavily based on the code in `mlpack/methods/sparse_autoencoder/sparse_autoencoder_function.hpp`; we could do some git wrangling to move that file instead of making a completely new one, and preserve the history that way (it'll take a little bit of work, but shouldn't be that hard).

File: src/mlpack/methods/ann/sparse_autoencoder_function.hpp [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+ *   const size_t numIterations = 100; // Maximum number of iterations.
+ *   const size_t numBasis = 10;
+ *   optimization::L_BFGS<SAEF> optimizer(encoderFunction, numBasis, numIterations);
+ *
+ *   arma::mat parameters = encoderFunction.GetInitialPoint();
+ *
+ *   // Train the model.
+ *   Timer::Start("sparse_autoencoder_optimization");
+ *   const double out = optimizer.Optimize(parameters);
+ *   Timer::Stop("sparse_autoencoder_optimization");
```
Comment (rcurtin) [inline src/mlpack/methods/ann/sparse_autoencoder_function.hpp]: The API here is the way Siddharth originally wrote it, using the standard mlpack optimizers to optimize the weights of the network.  But maybe it would be a better idea to make this class work with the `Trainer` class in `src/mlpack/methods/ann/`?  Kind of like the examples in `convolutional_network_test.cpp` and `feedforward_network_test.cpp`:

```
SAEF autoencoder(...);
Trainer<SAEF> trainer(autoencoder, ...);
trainer.Train(...);
```

This would help make all of the ANN-related code in mlpack have a unified interface.  I'd be interested in zoq's comments on this too, since I don't know his plans for what the API there will eventually look like.

**Comment by stereomatchingkiss** [comment]
>can I get a little more information on how you calculated that?
Of course, the file sparse_autoencoder_function already provide the example, I hope this is a correct way to measure the performance

>This file looks like it's heavily based on the code in mlpack/methods/sparse_autoencoder/sparse_autoencoder_function.hpp
Yes, I put it in ann mainly because this class break the old api and SparseAutoEncoder looks looks like part of the family member of neural network(I am not an expert of neural network, please correct me if I am wrong), for backward compatibility I do not change the original version. 

> But maybe it would be a better idea to make this class work with the Trainer class in src/mlpack/methods/ann/
I do not know how to make it work yet, but if this could work, it should be able to provide unified interface. I think I would wait for the comments of zoq's.

File: src/mlpack/methods/ann/sparse_autoencoder_function.hpp (hunk 3 above)
Comment (zoq) [inline src/mlpack/methods/ann/sparse_autoencoder_function.hpp]: I would go with the trainer class. The trainer class is not bounded with the network architecture, e.g. we can put a bias term in the first layer or in the first and second layer, etc. We can also eventually store the results and keep track of the optimization process using arbitrary performance functions.

File: src/mlpack/methods/ann/activation_functions/lazy_logistic_function.hpp [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+  /**
+   * Computes the logistic function without checking overflow or underflow
+   *
+   * @param x Input data.
+   * @param y The resulting output activation.
+   */
+  template<typename InputVecType, typename OutputVecType>
+  static void fn(const InputVecType& x, OutputVecType& y)
+  {
+    y = (1.0 / (1 + arma::exp(-x)));
```
Comment (zoq) [inline src/mlpack/methods/ann/activation_functions/lazy_logistic_function.hpp]: Originally my intention for the overflow checks in the LogisticFunction class was to avoid strange issues during the training process. But I'm intended to remove the checks in the LogisticFunction class, if you agree we can remove the LazyLogisticFunction class and just use the LogisticFunction class.

Comment (stereomatchingkiss) [inline src/mlpack/methods/ann/activation_functions/lazy_logistic_function.hpp]: I am totally fine with that

**Comment by stereomatchingkiss** [comment]
After some experiments, not all of the layers of ann are suitable for SparseAutoencoder, it need to do some modification(ex : you cannot just call the dropout function, you need to call the activation function like sigmoid or RELU after calling Forward), I would like to open two new folders

1 : SparseAutoencoder/layers
2 : SparseAutoencoder/activation_functions
to collect the useable layers and activation functions, the implementation will based on the layers and activation functions of ann

Besides, I think the function "GetNewFeatures" of SparseAutoencoder should call the activation function of the hidden layer rather the Sigmoid.

I will remove LazyLogisticFunction and remove the check of LogisticFunction as zoq mentioned

Any suggestions?

**Comment by zoq** [comment]
Maybe you can provide more information about the problem with the existing layers? I'm not sure I completely understand the problem.

I don't like the idea duplicating code, maybe we can workaround the problem or change the existing layer interface. I guess it wouldn't be a huge problem if we change the layers, so that they can be initialized with an activation function.

**Comment by stereomatchingkiss** [comment]
>Maybe you can provide more information about the problem with the existing layers?

No problem

1 : Not all of the layers are meaningful for SparseAutoencoder, like PoolingLayer, ConvolutionLayer, multiclass_classification_layer and so on. To make the users life easier, my initial idea is collect meaningful layers at another folder, this could reduce the chances of specify wrong layer type.

pros : 
a : The users would not need to bother about which layer is useable or scratch their head to understand the error messages generated at compile time or runtime
b : The layer could become simpler, since the member functions and data member can designed specifically for the SparseAutoencoder

cons : 
a : Generate duplicate codes

Another solution is using the type traits and static_assert to tell the users "This layer do not support by SparseAutoencoder".

    static_assert(IsValidLayerOfSE<HiddenLayer>, "this hidden layer do not support by SparseAutoencoder");
    static_assert(IsValidLayerOfSE<OutputLayer>, "this output layer do not support by SparseAutoencoder");

pros : 
a : reduce duplicate codes, you do not need to update two kind of similar layers 

cons : 
a :  the users need to deal with the compile time error messages, but I think is should be easy to read with the help of static_assert
b : the layer is a little bit complicated

2 : To make the dropout layer work, we need to add one more template parameter

    template <
        typename ActivationFunction = LogisticFunction,
        typename InputDataType = arma::mat,
        typename OutputDataType = arma::mat
    >
    class DropoutLayer{}

and call the activation after in the Forward function

    template<typename eT>
    void Forward(const arma::Mat<eT>& input, arma::Mat<eT>& output)
    {    
        // Scale with input / (1 - ratio) and set values to zero with probability
        // ratio.
        scale = 1.0 / (1.0 - ratio);      
        mask = arma::randu<arma::Mat<eT> >(input.n_rows, input.n_cols);
        mask.transform( [&](double val) { return (val > ratio); } );
        //following part could use tag dispatch to optimize, when the ActivationFunction
        //equal to EmptyFunction, it should called an empty function
        dropoutInput = input % mask * scale;
        ActivationFunction::fn(dropoutInput, output);
     }    

3 : The other problems of the is that the layout of the input is different with the SparseAutoencoder

SparseAutoencoder assume rows are features, cols are samples; but ann assume one arma::Mat<T> as a sample, and the whole 2D dimension of the arma::Mat<T> as features. However, DropoutLayer and BaseLayer do not affect by this.

Any suggestions?

**Comment by stereomatchingkiss** [comment]
Another problem of making the layers templatize is how could the users modify the value of the layers?
I would like to allow the users pass in the layers by constructor.

    SparseAutoencoder(const arma::mat& data,
                      size_t visibleSize,
                      size_t hiddenSize,
                      double lambda = 0.0001,
                      double beta = 3,
                      double rho = 0.01,
                      HiddenLayer &&hiddenLayer = HiddenLayer(),
                      OutputLayer &&outputLayer = OutputLayer());

This api do not looks good since there are too many parameters, it reminds me of legacy c codes or c with classes. Another constructor maybe a better choice

     SparseAutoencoder(const arma::mat& data,
                       size_t visibleSize,
                       size_t hiddenSize,
                       HiddenLayer &&hiddenLayer = HiddenLayer(),
                       OutputLayer &&outputLayer = OutputLayer());

I use && because some layer maybe expensive to copy, this way the users can only pass in the HiddenLayer and OutputLayer by rvalue, then the constructor will be able to move them. Let the SparseAutoencoder keep the layers could make codes easier to deal with(the users do not need to create another variables to keep the layers alive, let SparseAutoencoder handle them).

**Comment by rcurtin** [comment]
Okay, now that #452 is done, let's consider this one. :)

I want to make sure I'm on the same page as what's going on here, so please correct anything I've written below:

The idea are to improve the speed of the sparse autoencoder and flexibility of the sparse autoencoder, by reusing the code from `src/mlpack/methods/ann/`, which is both faster and allows the user to use different hidden layer types (like something other than the usual sigmoid layer).  Using a separate class called `SparseAutoencoder` could help keep users from getting confused with the tons of different options available with the ANN code.

I think that's a kind of accurate summary of what you're trying to do; if not, please correct it. :)  I am in support of this; for me, the primary two important things here are that (1) we can improve the speed of the `SparseAutoencoder` and (2) we can reuse the ANN code and remove virtually all of the existing code for `SparseAutoencoder`.  It would also be very nice that we could get more flexibility.

Before I start digging too much, do you think we could use the ANN code entirely to represent the sparse autoencoder?  i.e. `SparseAutoencoder` is actually a typedef of `ann::FFN<...>` (I have no idea what exactly the parameters to FFN would be; I haven't thought it through yet).  If that's the case, I think we should move that direction, get rid of the old `SparseAutoencoder` code, and then find ways to clean up the code and make it do the other things you want it to.

(If this isn't reasonable, let me know---I'm just trying to figure out a good way to integrate this code in a way that keeps the complexity of the resulting code down as much as possible.  I'm not a particular expert on autoencoders, though I am familiar enough with them, so my suggestions may not make 100% sense.  I like finding where code is duplicated and removing it. :))

**Comment by stereomatchingkiss** [comment]
>(1) we can improve the speed of the SparseAutoencoder

Yeah, I would like to do that, and intent to provide gpu version in the future.

>(2) we can reuse the ANN code and remove virtually all of the existing code for SparseAutoencoder

I would like to resue the ANN code + existing code for SparseAutoencoder if possible:)

>do you think we could use the ANN code entirely to represent the sparse autoencoder

I think it would not be able to do that, I wrote the reasons [at here!](https://github.com/mlpack/mlpack/pull/451#issuecomment-150390803). Either the layer of ANN need to do some changes and apply type_traits + static_assert to tell the users "this layer is a bad choice" or create a new folder to collect meaningful layer for autoencoder(but this would generate duplicate codes).

>do you think we could use the ANN code entirely to represent the sparse autoencoder

About the FNN part, I am not sure it could be resue or not, but it looks a lot similar with current implementation, I will need to dig into the codes, run some examples to find out the results are correct or not and compare the speed with old implementation. 

If it works, then we maybe able to provide autoencoder and sparse autoencoeder at the same time.

> I'm not a particular expert on autoencoders

Me neither, I am not even an expert of machine learning, I only know some basic ideas of it and need some decent tools to help me finish image recognition/classification tasks. At first I intent to write a library for myself, but later on I find out this is impractical and prefer to join an existing open source project:).

In short, I think expert like zoq would know the direction better.

**Comment by stereomatchingkiss** [comment]
I take a look at FNN, If I am correct, it is possible to implement autoencoder is just a simple feed forward network which try to reconstruct the input value. In this case, autoencoder and sparse autoencoder should be able to implement by FNN(correct me if I am wrong). I will try to implement one by FNN on this week.

**Comment by stereomatchingkiss** [comment]
Using FNN to implement autoencoder/sparse autoencoder remind me that most(or all) of the ann module do not support serialize yet, should I open another pull request to discuss the implementation details of serialization?

**Comment by rcurtin** [comment]
I think serialization for the ANN code would be great, but I don't know if @zoq has any plans for it already.

To the best of my understanding, you should be able to implement a sparse autoencoder as an FNN just like you said.

> I think it would not be able to do that, I wrote the reasons at here!. Either the layer of ANN need to do some changes and apply type_traits + static_assert to tell the users "this layer is a bad choice" or create a new folder to collect meaningful layer for autoencoder(but this would generate duplicate codes).

I remember you writing that, I just wanted to make sure that was the only reason.  I think there are a couple ways to go here---type_traits with a static_assert could be a good way to warn users, or, you can provide template typedefs so that the user is generally not building their own autoencoder, but instead just using the pre-existing `SparseAutoencoder` typedef.  i.e.

```
using SparseAutoencoder = FNN<...>;
using DenoisingAutoencoder = FNN<...>; // with the FNN class, maybe other types of autoencoders are easy too and can be provided in the future
```

Do you think that's a better solution?  This allows regular users the flexibility to do whatever they want by defining their own `FNN<...>` for a custom autoencoder, and users who just want to use a standard sparse autoencoder can use the `SparseAutoencoder` typedef (or whatever other types of autoencoders end up being available).

**Comment by zoq** [comment]
Serialization for the ANN code would be absolutely great. In fact, someone asks me some time ago, to implement a feature to save the trained model, so I think the serialization function will come in handy. Opening another pull request to discuss the implementation details sounds great. In most cases it shouldn't be more than saving the weights.


Reusing the existing ann code to implement an autoencoder should be straightforward. We have to implement a new performance function to calculate the reconstruction error, the regularization cost and the KL divergence cost terms to monitor the training process and because we don't have a target value we use the already provided input.

I don't have a good solution for warning the user if the layer combinations doesn't make sense at all. As @rcurtin already said using static_assert could be a good way to solve the problem, but I'm not sure if it's the best. Every time someone implements a new layer, he needs to think about in which context this layer makes sense, maybe this effort is negligible. So maybe it's enough, if we just define some default models. If a user is interested in a special layer combination, I would say he knows what combination makes sense. I guess this is what @rcurtin suggested as another solution.

**Comment by stereomatchingkiss** [comment]
>Opening another pull request to discuss the implementation details sounds great.

Then I would open one after I finished autoencoder(if no one open it yet)

>but instead just using the pre-existing SparseAutoencoder typedef

I think this is a good idea and prefer to do it this way

By the way, this is the layers(example 1) I try to feed into the FNN, do you have any suggestions?

    const arma::mat trainData = arma::randu<arma::mat>(4,4);
    const size_t visibleSize = trainData.n_rows;
    const size_t hiddenSize = trainData.n_rows / 2;
    const double range = sqrt(6) / sqrt(visibleSize + hiddenSize + 1);

    LinearLayer<RMSPROP, RandomInitialization>
            hiddenLayer(visibleSize, hiddenSize, {-range, range});
    BiasLayer<RMSPROP, RandomInitialization> hiddenBiasLayer(hiddenSize);
    BaseLayer<LogisticFunction> hiddenBaseLayer;

    LinearLayer<RMSPROP, RandomInitialization>
            outputLayer(hiddenSize, visibleSize, {-range, range});
    BiasLayer<RMSPROP, RandomInitialization> outputBiasLayer(hiddenSize);
    BaseLayer<LogisticFunction> outputBaseLayer;

    auto network = std::tie(hiddenLayer, hiddenBiasLayer, hiddenBaseLayer,
                            outputLayer, outputBiasLayer, outputBaseLayer);
    auto resultLayer = OneHotLayer();
    FFN<decltype(network), OneHotLayer, SparseErrorFunction>
            ffn(network, resultLayer);

>We have to implement a new performance function to calculate the reconstruction error

This one may need to change the implementation details of the ann module.

There are two problems need to be solved

1 : The performance function of ann cannot store the condition by current implementation
2 : The performance function cannot access w1 and w2 of the network

    template<typename DataType, typename ErrorType, typename... Tp>
    double OutputError(const DataType& target,
                       ErrorType& error,
                       const std::tuple<Tp...>& t)
    {
      // Calculate and store the output error.
      outputLayer.CalculateError(
          std::get<sizeof...(Tp) - 1>(t).OutputParameter(), target, error);

      // Masures the network's performance with the specified performance
      // function.
      return PerformanceFunction::Error(
          std::get<sizeof...(Tp) - 1>(t).OutputParameter(), target);
    }

As you see, what I could get from the Error api is the target(in autoencoder, it is the original training data) and the activation of the output layer(please refer to example 1), but not the w1 and w2 of the hidden layer and the output layer.But we need them to calculate the weightDecay. Besides, I guess the users may want to setup the weight of KL divergence and weightDecay too.

If we could store the PerformanceFunction, this problem could be solved.Rather than calling static Error function, call it like 

    performanceFunction.Error(std::get<sizeof...(Tp) - 1>(t).OutputParameter(), target);

change the constructor to

    FFN(const LayerTypes& network, OutputLayerType& outputLayer, 
          PerformanceFunction performanceFunction = CrossEntropyErrorFunction())
        : network(network), outputLayer(outputLayer), trainError(0),
    performance(std::forward<PerformanceFunction>(performanceFunction))
      {
        // Nothing to do here.
      }


And call it like

    const double beta = 0.3; //weight of KL divergence
    const double lambda = 0.001; //weight of weight decay
    FFN<decltype(network), OneHotLayer, SparseErrorFunction>
    ffn(network, resultLayer, 
        SparseErrorFunction(hiddenLayer.Weights(), outputLayer.Weights(), beta, lambda));


I think this should be more flexible. The other question is, do you have any plans to provide move constructor and move assignments for ann modules? Most of the class would not be needed to change anything if armadillo already implement it, maybe a better way is ask armadillo implement one, have no idea why they do not provide it.

**Comment by stereomatchingkiss** [comment]
Ok, find out the problem, we need a new "LinearLayer" which customize for the SparseAutoencoder, the problem is this part

    arma::mat klDivGrad = beta * (-(rho / rhoCap) + (1 - rho) / (1 - rhoCap));
    klDivGrad.elem(arma::find_nonfinite(klDivGrad)).zeros();
    diff2 = parameters.submat(l1, 0, l3 - 1, l2 - 1) * delOut +
              arma::repmat(klDivGrad, 1, data.n_cols);
    arma::mat delHid;
    hiddenLayerFunc.Backward(hiddenLayer, diff2, delHid);

We would need to have do some special calculation if this is sparse autoencoder, first we need the rho, second we need the rhoCap from the performance function, other parts are done by FNN already(please correct me if I am wrong). Good news are we do not need to modify the codes ann.

    LinearLayer<RMSPROP, RandomInitialization>
            hiddenLayer(visibleSize, hiddenSize, rho, {-range, range});  

The performance function should be able to accept the reference of RhoCap of hiddenLayer

    SparseErrorFunction(hiddenLayer.Weights(), outputLayer.Weights(), hiddenLayer.RhoCap(), 
                        beta, lambda)

Any suggestion?

**Comment by zoq** [comment]
>If we could store the PerformanceFunction, this problem could be solved.Rather than calling static Error function, call it like``

You are right. The suggested change looks great to me.


>The other question is, do you have any plans to provide move constructor and move assignments for ann modules? ``

I haven't had the time yet to provide a move constructor, but it's on my list. If you like, to provide some code, I'll be happy to merge the changes. But don't feel obligated.

>The performance function should be able to accept the reference of RhoCap of hiddenLayer, because both of the hiddenLayer and the performance function need to use it.

If we pass the Weights and the RhoCap parameter we have to handle the pass of the parameter inside of the fnn class, if we pass an instance of the network, we can handle the parameter inside the performance function.

If we implement a getter function for the ``RhoCap`` parameter inside the new ``SparseLinearLayer``, we don't need to provide a special constructor. We can update the parameter inside the error function once we call the performance function. What do you think about the idea?

**Comment by stereomatchingkiss** [comment]
>If we pass the Weights and the RhoCap parameter we have to handle the pass of the parameter inside of the fnn class

Do you mean rather than

    SparseErrorFunction(hiddenLayer.Weights(), outputLayer.Weights(), hiddenLayer.RhoCap(), 
                        beta, lambda)

Pass whole network into it would be better?

    SparseErrorFunction(network, beta, lambda)

I have no problem with this one, if you think it is ok I would start to implement it.

>If you like, to provide some code, I'll be happy to merge the changes. But don't feel obligated.

Ok, would try to implement it if I got time

**Comment by zoq** [comment]
Ah I missed the beta and lambda parameter, I guess in this case we should go with your suggested interface. I thought we could extract every information from the network, which would result in a single parameter we have to pass to the performance function.

>Ok, would try to implement it if I got time

Sounds, great, I do the same if I have the time.

**Comment by stereomatchingkiss** [comment]
>which would result in a single parameter we have to pass to the performance function.

I think both designs have their pros and cons

Implementing SparseCost function, and I find out the Error function need one more parameter--error

    double Error(const DataType&, const DataType&, const DataType &error)

The problem is FNN call the function as following

    return performanceFunc.Error(
            std::get<sizeof...(Tp) - 1>(t).OutputParameter(), target);

But I need the error parameter

    return performanceFunc.Error(
            std::get<sizeof...(Tp) - 1>(t).OutputParameter(), target, error);



Now there are two solutions in my mind 

1 : make all of the Error function accept one more parameter
2 : Use SFINAE to determine which function to called

a : declare a traits 
    HAS_MEM_FUNC(Error, HasErrorCheck);
b : use SFINAE to select appropriate function at compile time

    typename std::enable_if<
          HasErrorCheck<T, double(T::*)(const U&, const U&, const U&)>::value, double>::type

I prefer solution 1, what is your opinion?

**Comment by zoq** [comment]
I prefer the first version. In fact, I've implemented the negative log likelihood for another model, which also requires the error value, so now there isn't just one function that requires an interface modification.

**Comment by stereomatchingkiss** [comment]
>I prefer the first version

Ok, then I would pick first version. There are many details need to deal with, you will find out some new layers and error function design for sparse autoencoder.

**Comment by stereomatchingkiss** [comment]
When implementing sparse autoencoder by FNN, I find out there are some copy could be eliminated, I think the first one should be deal with is the Forward function.

    template<size_t I = 0, typename DataType, typename... Tp>
    void Forward(const DataType& input, std::tuple<Tp...>& t)
    {
      std::get<I>(t).InputParameter() = input;    
      //.......
    }

The input data could be quite big(several MB or GB), it would be better if we could eliminate this copy. If there are another api like

    std::get<I>(t).InputParameter(input);

Then it is possible to use a pointer point to the input if we know the input would not modified, but this would 
1 : add one more api to the layers
2 : part of the layers may need another template parameter to determine the input parameter is const or not(I am not sure about this part).

**Comment by stereomatchingkiss** [comment]
About the part of prediction, if I could access underlying network, it would be easier

    using ActivateFunction = typename HiddenLayer::ActivateFunction;

    arma::mat const encodeInput = w1 * input+ arma::repmat(b1, 1, input.n_cols);
    ActivateFunction::fn(encodeInput , features);

As you can see, to extract the features from the input, I do not need to forward the whole layers, I would need to access underlying network of the FFN, could I add a public api to access the network?

**Comment by stereomatchingkiss** [comment]
The Error function of the performance function need the network parameters,because the SparseCost need to access the Weights and RhoCap

The FFN take the copy of the network but not the reference of it, so it is no use to use the pointer to point to the weights and RhoCap of the network

      //in this case, input can get from the network too
      double Error(const DataType &target, const DataType &error, const Network &network)
      {
        // Calculate squared L2-norms of w1 and w2.
        const double wL2SquaredNorm =
            arma::accu(std::get<0>(network).Weights() % std::get<0>(network).Weights()) +
            arma::accu(std::get<3>(network).Weights() % std::get<3>(network).Weights());    
  
        const double sumOfSquaresError =
            0.5 * arma::accu(error % error) / target.n_cols;
        const double weightDecay = 0.5 * lambda * wL2SquaredNorm;
        const double klDivergence =
            beta * arma::accu(rho * arma::trunc_log(rho / std::get<0>(network).RhoCap()) + (1 - rho) *
                              arma::trunc_log((1 - rho) / (1 - std::get<0>(network).RhoCap())));    
    
        // The cost is the sum of the terms calculated above.
        return sumOfSquaresError + weightDecay + klDivergence;
      }

If you agree with it, I would make the necessary changes.

**Comment by zoq** [comment]
> If you agree with it, I would make the changes.

I agree, using the network as a single parameter simplifies the parameters of the error function. Additionally, if someone needs more parameters in the future, the parameters can be extracted from the network.

> As you can see, to extract the features from the input, I do not need to forward the whole layers, I would need to access underlying network of the FFN, could I add a public api to access the network?

I don't see any convincing argument against a public function to access the network. So please go ahead.

> When implementing sparse autoencoder by FNN, I find out there are some copy could be eliminated, I think the first one should be deal with is the Forward function. [...] I would like to deal with this issue after this pull request done, any idea?

Yeah, let us discuss the issue in another thread. I guess we can change the current API, as you suggested and remove the current getter function.

**Comment by stereomatchingkiss** [comment]
I upload the sparse_autoencoder, which place at ann/autoencoder/sparse_autoencoder.hpp

I haven't test it with the Trainer class and write the unit test yet, but the value obtain by FeedForward, FeedBackWard  are same as the old class.

If you think there are things need to be changed(api, name, place the files at different folder etc), please give me some opinions, I will move on to next step after you think this design is ok

**Comment by stereomatchingkiss** [comment]
Finally, I can run the Train function without crash, but the training results is correct or good enough or not is another question, need more test.

Sample codes : 

    const arma::mat trainData = arma::randu<arma::mat>(2,4);
    const size_t visibleSize = trainData.n_rows;
    const size_t hiddenSize = trainData.n_rows / 2;    

    using AnnSae = ann::LogisticSparseAutoencoder<>;
    size_t const batch_size = trainData.n_cols;
    size_t const epoch = 100;
    AnnSae  sae(visibleSize, hiddenSize, batch_size);
    sae.Train(trainData, epoch);

**Comment by zoq** [comment]
Sorry for the slow response, the code looks good to me. I'll make a few minor comments and formatting fixes once the code is merged. I'll fix that separately, no need to worry about it here.

I guess since the forward and backward functions provide the same output as the old code, the test should run without any problems. If that's the case I'll be happy to merge the changes.

**Comment by stereomatchingkiss** [comment]
>I guess since the forward and backward functions provide the same output as the old code, the test should run without any problems. If that's the case I'll be happy to merge the changes.

If you think so, then I would write a simple test case to test the output of forward and backward later on. After this has been done, we could deal with another problem

**Comment by zoq** [comment]
No need to write another test for the forward and backward function, if the existing sparse autoencoder test works with the new code it's absolutely fine. I guess ```sparse_autoencoder_test_2.cpp``` is the modified test and we can remove the ```sparse_autoencoder_test.cpp``` file and rename the other test.

**Comment by stereomatchingkiss** [comment]
>I guess sparse_autoencoder_test_2.cpp is the modified test and we can remove the sparse_autoencoder_test.cpp file and rename the other test.

I would like to do that too, but the problem is old sparseAutoencoder and the new one do not have the same api, so the tests need to have some change.

**Comment by stereomatchingkiss** [comment]
I upload the test file, the name is "sparse_autoencoder_test_3.cpp". 
If you think it is ok, you could rename it to "sparse_autoencoder_test".

The api of the FFN are different than the original sparseAutoencoder, therefore
we need to do some change on the original codes.

**Comment by stereomatchingkiss** [comment]
Finally, all jobs have been done. I do some change on FFN(do not take the reference of outputLayer but add a public api to users, this make serialization become easier).

I use FeedForward to get the Evaluate value(Evaluate api do not give the value)
Call FeedForward and FeedBackWard to get the gradient value

Please check the test cases, they test with the same data(except of the gradient part) but with different api, I hope my implementation meet the requirement of the semantic. 

If you think everything going well, you can merge it now. Thanks for taking time review

**Comment by rcurtin** [comment]
When the merge is done, I'll go ahead and add a reverse-compatibility layer for the 2.x.x releases and then remove the old SparseAutoencoder code.

**Comment by zoq** [comment]
Thanks for the contribution. I made a couple of changes:

- moved the main sparse autoencoder into a separate folder in 4ad39f8651da0
- minor formatting and comment fixes in 896937d03c5eb32a4e980
- modified the sparse autoencoder test in 443ecdcd35d, so that it uses the SparseAutoencoder class you already implemented. Since we already have a gradient test for each activation function, I removed the gradient sparse autoencoder test.

Let me know if I messed anything up.

Since, we have this nice SparseAutoencoder class, it should be easy provide a reverse-compatibility layer for the 2.x.x releases. I'll go and write the necessary code, if nobody really likes to do it. 

We should also think about a test case that tests the code in combination with an optimizer. I run into a couple of problems, once I tested the code with the existing trainer class. I solved the issues in f34ae33e2ccdaca68dc7. Another test could also test the ability to work with additional layer. We only test the standard sparse autoencoder model structure (input layer, hidden layer, output layer). Wich is fine since the former code uses this static model structure, but since we build the sparse autoencoder using the ann modules, we have the ability to add a bunch of interesting layer, e.g. Dropout layer.

**Comment by zoq** [comment]
Also, I think a command line program would be neat feature.

**Comment by stereomatchingkiss** [comment]
Thanks for the fix.

I would provide a command line programs, with different activations(relu, tanh, sigmoid) and dropout(if it works) after serialization of FFN done.
